### PR TITLE
feat(replica): add debug command - replica offset

### PR DESCRIPTION
**Pull request by adiholden**
Signed-off-by: adi_holden <adi@dragonflydb.io>

 1. Add DEBUG REPLICA OFFSET command which will return the sync id and array offsets for each flow
 2. fix bug in Replica::TransactionReader::NextTxData

### CODE REVIEW

File: src/server/debugcmd.cc [hunk 1]
```diff
@@ -104,6 +104,8 @@ void DebugCmd::Run(CmdArgList args) {
         "      existing RDB file.",
         "REPLICA PAUSE/RESUME",
         "    Stops replica from reconnecting to master, or resumes",
+        "REPLICA OFFSET",
```
Comment (romange) [inline src/server/debugcmd.cc:107]: mention that you send a sync id together with the array of offsets for each replication flow.

File: src/server/replica.cc [hunk 2]
```diff
@@ -1201,29 +1220,30 @@ bool Replica::TransactionData::IsGlobalCmd() const {
...
+bool Replica::TransactionReader::ReturnEntryOOO(const TransactionData& tx_data,
+                                                const journal::ParsedEntry& entry) {
+  return !tx_data.commands.empty() && entry.opcode == journal::Op::EXPIRED &&
+         tx_data.dbid != entry.dbid;
 }
 
 auto Replica::TransactionReader::NextTxData(JournalReader* reader, Context* cntx)
     -> optional<TransactionData> {
   io::Result<journal::ParsedEntry> res;
+  TransactionData next_tx;
```
Comment (adiholden) [inline src/server/replica.cc:1232]: I fixed the flow in this function, as std::move does not guarantees to empty the data (line 1246) from struct, we need to use swap

File: src/server/server_family.cc [hunk 3]
```diff
@@ -755,6 +755,18 @@ void ServerFamily::PauseReplication(bool pause) {
   }
 }
 
+std::optional<ReplicaOffsetInfo> ServerFamily::GetReplicaOffsetInfo() {
+  unique_lock lk(replicaof_mu_);
+
+  // Switch to primary mode.
+  if (!ServerState::tlocal()->is_master) {
+    auto repl_ptr = replica_;
+    CHECK(repl_ptr);
```
Comment (dranikpg) [inline src/server/server_family.cc:764]: So we just crash if we're neither replica nor master ? 🤔

File: src/server/debugcmd.cc [hunk 4]
```diff
@@ -190,6 +192,19 @@ void DebugCmd::Replica(CmdArgList args) {
   if (opt == "PAUSE" || opt == "RESUME") {
     sf_.PauseReplication(opt == "PAUSE");
     return (*cntx_)->SendOk();
+  } else if (opt == "OFFSET") {
+    const auto& offset_info = sf_.GetReplicaOffsetInfo();
```
Comment (dranikpg) [inline src/server/debugcmd.cc:196]: the reference is valid here, but its better to take by value

File: src/server/server_family.cc (hunk 3 above)
Comment (adiholden) [reply src/server/server_family.cc:764]: this is the same check taken from ServerFamily::ReplicaOf and in ServerFamily::PauseReplication
if server is replica we expect replica_ to be valid.

**Review by dranikpg — DISMISSED** [review]

**Review by dranikpg — APPROVED** [review]

**Pull request merged**